### ISSUE TITLE
Additional info in feature, scenario and steps

### ISSUE THREAD

**Issue by tuscias** [LightBDD/LightBDD#191]
#### Description
Hi, 
i'm implementing my own scenario decorator and noticed that scenario object doesn't have info about his parent, i.e feature. This information is needed to create Allure reports, where i have to specify a feature id for each scenario so they would link correctly:

public class ScenarioAllureDecorator : IScenarioDecorator
    {
        private static readonly AllureLifecycle allure = AllureLifecycle.Instance;

        public async Task ExecuteAsync(IScenario scenario, Func<Task> scenarioInvocation)
        {

            var scenarioResult = new TestResult
            {
                uuid = Guid.NewGuid().ToString(),
                name = scenario.Info.Name.ToString(),
                fullName = "FULL NAME",
                status = Status.passed,
                labels = new List<Label>
                {
                    Label.Thread(),
                    Label.Host(),
                    Label.Feature("Dummy"), // <---- HERE: how to get feature name
                    Label.Story(scenario.Info.Name.ToString()),
                }
            };
            allure.StartTestCase(scenarioResult);
            try
            {
                await scenarioInvocation();
            }
            catch (Exception ex)
            {
                allure.StopTestCase(scenarioResult.uuid);
                allure.WriteTestCase(scenarioResult.uuid);
                throw ex;
            }

        }
    }

Am i missing something? Is this could approached a bit differently?

**Comment by Suremaker** [comment]
Hello @tuscias ,

At this moment the child items do not have the ability to track their parent (scenario result/info cannot retrieve the feature result/info). I will make a feature for obtaining it.

If you just want to generate the report after all tests are done, you can achieve it by making own report writer. Here is a wiki page on [generating reports](https://github.com/LightBDD/LightBDD/wiki/Generating-Reports).

What you could do is to make own report writer:
```c#
class MyEmailingReporter : IReportWriter
{
    public void Save(params IFeatureResult[] results)
    {
        foreach (var featureResult in results)
        {
            // report feature with all it's scenarios and steps
        }
    }
}
```

The `IFeatureResult` contains the feature details and its scenarios with scenario info as well as execution results, timings etc.

...then register your writer in LightBDD:
```c#
protected override void OnConfigure(LightBddConfiguration configuration)
{
    configuration
        .ReportWritersConfiguration()
        .Add(new AllureReporter());
}
```

Please note that the report will be generated after all tests are done, so you won't be able to implement any execution notifications as the tests are running this way...

----
If you are interested in the immediate notification of the test progress, I would say the valid method would be to implement own [progress notifier](https://github.com/LightBDD/LightBDD/wiki/Test-Progress-Notification#customizing-progress-notification) and update report as the tests are executing:

```c#
public class AllureProgressNotifier : IScenarioProgressNotifier, IFeatureProgressNotifier
{
    // code updating reports as tests progresses
}
```

then the registration could be done like that:

```c#
var allureNotifier = new AllureProgressNotifier();
configuration.Get<FeatureProgressNotifierConfiguration>()
	.AppendNotifiers(allureNotifier);

configuration.Get<ScenarioProgressNotifierConfiguration>()
	.AppendNotifierProviders(() => allureNotifier);
```

but with this method, you will get to the same problem as with the decorator where `void NotifyScenarioFinished(IScenarioResult scenario)` won't have access to the feature, which would mean that you could only retrieve all the data in `void NotifyFeatureFinished(IFeatureResult feature)`.

---

As mentioned at the beginning of the response, I will have to implement the parent tracking for obtaining incremental/interactive report generation (please let me know if that is your case).

If you just want to make a report at the end, I would suggest going with option 1 as it's the simplest one.

I hope it helps,
Cheers.

**Comment by Suremaker** [comment]
Hey @tuscias ,

Please take a look at PR #192 
The `IScenarioInfo` and `IStepInfo` has been updated with `Parent` property that refers to parent metadata.
Moreover the `IFeatureInfo`, `IScenarioInfo` and `IStepInfo` got `RuntimeId` that is an unique runtime identifier of the metadata.

Please let me know if the change is sufficient for you to continue with your integration. I have a plan to merge and release it on Monday. 

Please also be aware that on Thursday I am going on holiday, which means I won't be able making any changes till the beginning of September.

**Comment by tuscias** [comment]
Yes, that is exactly how I wanted it to work!
Thanks for super fast PR!

**Comment by Suremaker** [comment]
Ok, I'll prepare the release tonight

**Comment by Suremaker** [comment]
Released with LightBDD 3.1.0
